clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1581,48 +1577,27 @@ private Fetch<K, V> collectFetch() {
      *
-     * @throws AuthenticationException If authentication fails. See the exception for more details
-     * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
-     *             defined
      * @return true iff the operation completed without timing out
+     * @throws AuthenticationException       If authentication fails. See the exception for more details
+     * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
+     *                                       defined
      */
     private boolean updateFetchPositions(final Timer timer) {
         try {
-            // Validate positions using the partition leader end offsets, to detect if any partition
-            // has been truncated due to a leader change. This will trigger an OffsetForLeaderEpoch
-            // request, retrieve the partition end offsets, and validate the current position against it.
-            applicationEventHandler.addAndGet(new ValidatePositionsEvent(calculateDeadlineMs(timer)));
-
-            cachedSubscriptionHasAllFetchPositions = subscriptions.hasAllFetchPositions();
-            if (cachedSubscriptionHasAllFetchPositions) return true;
-
-            // Reset positions using committed offsets retrieved from the group coordinator, for any
-            // partitions which do not have a valid position and are not awaiting reset. This will
-            // trigger an OffsetFetch request and update positions with the offsets retrieved. This
-            // will only do a coordinator lookup if there are partitions which have missing
-            // positions, so a consumer with manually assigned partitions can avoid a coordinator
-            // dependence by always ensuring that assigned partitions have an initial position.
-            if (isCommittedOffsetsManagementEnabled() && !initWithCommittedOffsetsIfNeeded(timer))
-                return false;
-
-            // If there are partitions still needing a position and a reset policy is defined,
-            // request reset using the default policy. If no reset strategy is defined and there
-            // are partitions with a missing position, then we will raise a NoOffsetForPartitionException exception.
-            subscriptions.resetInitializingPositions();
-
-            // Reset positions using partition offsets retrieved from the leader, for any partitions
-            // which are awaiting reset. This will trigger a ListOffset request, retrieve the
-            // partition offsets according to the strategy (ex. earliest, latest), and update the
-            // positions.
-            applicationEventHandler.addAndGet(new ResetPositionsEvent(calculateDeadlineMs(timer)));
-            return true;
+            UpdateFetchPositionsEvent updateFetchPositionsEvent = new UpdateFetchPositionsEvent(calculateDeadlineMs(timer));
```

Comment:
I believe that this will create an `UpdateFetchPositionsEvent` for every iteration of the inner loop in `AsyncKafkaConsumer.poll(Duration)`. Previously, the `cachedSubscriptionsHasAllFetchPositions` was an attempt at optimising this area, which was probably not entirely satisfactory but it was well intentioned. Generally, once the positions have been validated, there's no need for any of this unless something nasty happens.  I feel there's an optimisation waiting here. We know when the set of assigned partitions changed. We also know when an error such as log truncation occurred. We know when an `UpdateFetchPositionsEvent` completed successfully. So, I think we probably can be a bit more deliberate about creating the event only when it's needed.

Reply:
Good point. The trick is we do need to generate an event on every iteration of the poll just because we need to go to the background thread (even if we end up finding that we have all the positions already). This is to:

1. Get errors that may have been found in the background on a previous update positions attempt
2. Trigger validate positions (we do it on every poll as the classic consumer, and the validate common logic  internally determines if it really needs to do something or not)
3. check `subscriptions.hasAllFetchPositions()` (this we want to do it in the background to avoid races on the assignment that it checks)

And I'll stop here, just to point out that this is what we expect will happen on every poll iteration, under happy/regular scenarios (all partitions have valid positions). This is what that `UpdateFetchPositionsEvent` is needed for, (not really updating anything, but doing what it needs to determine that it doesn't need to). It's probably more of a MaybeUpdateFetchPositionsEvents (weird or better?)

Then, to cater for the scenarios where something fails/changes, the same event gets more interesting. If it does not "hasAllFetchPositions", it will attempt to retrieve committed offsets or partition offsets as needed, the meaty part. 

This is the reasoning behind having an event on every poll, what do you think?

Reply:
Sounds good to me. Thanks for explanation. We're well on the way to removing the SubscriptionState references from the application thread entirely.